### PR TITLE
[application] Transfer learning using just C-API

### LINKED ISSUE #451: Create a transfer learning application
Create a transfer learning application with C-API for reference to our customers, as this will be the primary use case of our library.

### PR DESCRIPTION
**Pull request by kparichay**
Update transfer learning application to use just C-API
This divides the application into 3 parts -
1. Running the fixed part of the model - done using nnstreamer-single C-API
2. Training the trianable part of the model - done with nntrainer C-API
3. Running the testing of the trainable part of the model - done with nnstreamer-pipeline C-API with nntrainer tensor filter extension

The outputs of section 1 is kept loaded in the memory than saving to disk and loading back

Commit - 2
Added callback for testing the model
The callback prints the top-1 predicted output for a given image file
Increased number of epochs back to 1000 to acheive better accuracy

Resolves #451 

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM. Nice work you have done. One thing I like to propose is that how about add some explanation in README. like code structures and how it works including the differences between tizen and other platform.

**Comment by kparichay** [comment]
> LGTM. Nice work you have done. One thing I like to propose is that how about add some explanation in README. like code structures and how it works including the differences between tizen and other platform.

I will add it in the README.md as a new PR today.

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp [hunk 1]
```diff
@@ -62,15 +68,122 @@
...
+      char *test_file_path = NULL;
+      status = asprintf(&test_file_path, "%s/%s/%s%d.bmp", data_path,
+                        label_names[i], label_names[i], j + 1);
+      if (status < 0 || test_file_path == NULL) {
+        status = -errno;
+        goto fail_exit;
+      }
+
+      count = i * NUM_DATA_PER_LABEL + j;
+      memset(label_data[count], 0, LABEL_SIZE * sizeof(float));
```
Comment (zhoonit) [inline Applications/TransferLearning/Draw_Classification/jni/main.cpp:171]: It seems that some architecture might not be using `all bit 0` as representing 0.0f. 
Check this out
https://stackoverflow.com/a/37555983

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp [hunk 2]
```diff
@@ -400,10 +554,18 @@ int testModel(std::string data_path, const TFLiteData &tflite_data,
 int main(int argc, char *argv[]) {
   int status = ML_ERROR_NONE;
   if (argc < 3) {
-    std::cout << "./TransferLearning Config.ini resources\n";
-    exit(0);
+#if defined(__TIZEN__)
+    ml_loge("./TransferLearning Config.ini resources.");
+#else
+    std::cout << "./TransferLearning Config.ini resources." << std::endl;
```
Comment (zhoonit) [inline Applications/TransferLearning/Draw_Classification/jni/main.cpp:560]: Could add information or checks about trailing slash in the resource directory?
```suggestion
    std::cout << "./TransferLearning Config.ini resources" << std::endl;
```

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp [hunk 3]
```diff
@@ -62,15 +68,122 @@
...
+  /* generate dummy data */
+  status = ml_tensors_data_create(in_res, &input);
+  if (status != ML_ERROR_NONE)
+    goto fail_in_info;
+
+  status = ml_tensors_data_get_tensor_data(input, 0, &raw_data, &data_size);
+  if (status != ML_ERROR_NONE)
+    goto fail_in_data;
+
+  in = tflite::label_image::read_bmp(test_file_path, inputDim, inputDim + 1,
```
[the remark this answers is not in the document]
Comment (kparichay) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:103]: yes, this is from bitmap header, with file existing in the folder.

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp (hunk 1 above)
Comment (kparichay) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:171]: done.

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp (hunk 2 above)
Comment (kparichay) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:560]: the trailing slash is handled internally by always adding slash at the end of resources path.

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:103]: I see. Thanks

File: Applications/TransferLearning/Draw_Classification/jni/main.cpp (hunk 2 above)
Comment (zhoonit) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:560]: This application used to require adding the last slash for the resource path. Check out `meson.build`.  btw. I guess the test argument should be modified as well.

`test('app_training', e, args: [join_paths(res_path, 'Training.ini'), res_path + '/'], timeout: 60)`

Comment (kparichay) [reply Applications/TransferLearning/Draw_Classification/jni/main.cpp:566]: updated to `test('app_training', e, args: [join_paths(res_path, 'Training.ini'), res_path], timeout: 60)`

**Review by zhoonit — APPROVED** [review]
LGTM

**Pull request merged**